types/wlr_pointer_gestures_v1.c

```diff
@@ -0,0 +1,335 @@
+	return wl_resource_get_user_data(resource);
+}
+
+void wlr_pointer_gestures_v1_send_swipe_begin(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_seat *seat,
+		uint32_t time_msec,
+		uint32_t fingers) {
+	struct wlr_surface *focus = seat->pointer_state.focused_surface;
+	if (focus == NULL) {
+		return;
+	}
+
+	struct wl_client *focus_client = wl_resource_get_client(focus->resource);
+	uint32_t serial = wl_display_next_serial(
+			wl_client_get_display(focus_client));
+
+	struct wl_resource *gesture;
+	wl_resource_for_each(gesture, &gestures->swipes) {
+		struct wlr_seat *gesture_seat = seat_from_pointer_resource(gesture);
+		struct wl_client *gesture_client = wl_resource_get_client(gesture);
+		if (gesture_seat != seat || gesture_client != focus_client) {
+			continue;
+		}
+		zwp_pointer_gesture_swipe_v1_send_begin(gesture, serial,
+				time_msec, focus->resource, fingers);
+	}
+}
+
+void wlr_pointer_gestures_v1_send_swipe_update(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_seat *seat,
+		uint32_t time_msec,
+		double dx,
+		double dy) {
+	struct wlr_surface *focus = seat->pointer_state.focused_surface;
+	if (focus == NULL) {
+		return;
+	}
+
+	struct wl_client *focus_client = wl_resource_get_client(focus->resource);
+
+	struct wl_resource *gesture;
+	wl_resource_for_each(gesture, &gestures->swipes) {
+		struct wlr_seat *gesture_seat = seat_from_pointer_resource(gesture);
+		struct wl_client *gesture_client = wl_resource_get_client(gesture);
+		if (gesture_seat != seat || gesture_client != focus_client) {
+			continue;
+		}
+		zwp_pointer_gesture_swipe_v1_send_update(gesture, time_msec,
+				wl_fixed_from_double(dx), wl_fixed_from_double(dy));
+	}
+}
+
+void wlr_pointer_gestures_v1_send_swipe_end(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_seat *seat,
+		uint32_t time_msec,
+		bool cancelled) {
+	struct wlr_surface *focus = seat->pointer_state.focused_surface;
+	if (focus == NULL) {
+		return;
+	}
+
+	struct wl_client *focus_client = wl_resource_get_client(focus->resource);
+	uint32_t serial = wl_display_next_serial(
+			wl_client_get_display(focus_client));
+
+	struct wl_resource *gesture;
+	wl_resource_for_each(gesture, &gestures->swipes) {
+		struct wlr_seat *gesture_seat = seat_from_pointer_resource(gesture);
+		struct wl_client *gesture_client = wl_resource_get_client(gesture);
+		if (gesture_seat != seat || gesture_client != focus_client) {
+			continue;
+		}
+		zwp_pointer_gesture_swipe_v1_send_end(gesture, serial,
+				time_msec, cancelled);
+	}
+}
+
+static const struct zwp_pointer_gesture_swipe_v1_interface swipe_impl = {
+	.destroy = resource_handle_destroy,
+};
+
+static void get_swipe_gesture(struct wl_client *client,
+		struct wl_resource *gestures_resource,
+		uint32_t id,
+		struct wl_resource *pointer_resource) {
+	struct wlr_seat_client *seat_client =
+		wlr_seat_client_from_pointer_resource(pointer_resource);
+
+	if (seat_client == NULL) {
+		wl_client_post_no_memory(client);
```

Comment:
The protocol doesn't define any errors, so idk what else to do here, just an error on the gestures_resource?

Reply:
oh, right.